src/v/utils/queue_depth_control.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+#include "seastarx.h"
+#include "utils/ema.h"
+
+#include <seastar/core/semaphore.hh>
+
+/*
+ * exponential moving average queue depth control. given a maximum latency
+ * update the number of units available in the contained semaphore which can be
+ * used to limit queue depth to control latency.
+ *
+ * when updated with a sample of `0` idleness is assumed. when idle if the depth
+ * falls below the idle depth then idle depth takes precedence. otherwise, the
+ * depth is constrained by min/max depth configuration parameters.
+ */
+class queue_depth_control {
+public:
+    queue_depth_control(
+      std::chrono::milliseconds max_latency,
+      double smoothing_factor,
+      size_t idle_depth,
+      size_t min_depth,
+      size_t max_depth)
+      : _max_latency(max_latency.count()) // NOLINT
+      , _gamma(smoothing_factor)
+      , _idle_depth(idle_depth)
+      , _min_depth(min_depth)
+      , _max_depth(max_depth)
+      , _curr_depth(_idle_depth)
+      , _queue(_curr_depth) {}
+
+    void update(const double sample) {
+        auto new_depth = static_cast<size_t>(
+          (1.0 - _gamma) * static_cast<double>(_curr_depth));
```

Comment:
given that this is executed on every produce, did we find a way to do integer math instead of floating point math ?

Reply:
it is not executed on every produce. this calculation occurs every 7 seconds (by default). only two integer additions occur on every produce.

Reply:
ohhh. i see.